SPM Prep - Convert "WordPress JSON date" logic from Objective-C to Swift

mokagio opened the pull request:
### Description

This logic will need to live in the "Core API" package and the Jetpack/WordPress specific objects. It could have been extracted into a dedicated package, but for the sake of simplicity, it's been converted to Swift so it can live in "Core API".

Notice that:

1. I first added unit tests (smoke tests, more like) to ensure consistent behavior between implementation
2. The Objective-C `dateWithWordPressComJSONString:` method was translated by the compiler into `NSDate(wordPressComJSONString:)`. In my conversion, I made it `.with(wordPressComJSONString:)`. I think it would have been possible to define a convenience initializer to keep the API the same, but the approach I took seemed cleaner at the time. The method is only used internally to WordPressKit (see commit details) so, while it might count as a breaking change, I'm tempted to ignore it. Let me know if you think I'm being slack, or if it sounds like a reasonable compromise to move ahead without distractions. 

### Testing Details

See the new unit tests.

### Next up

Basically, cherry-pick the work from https://github.com/wordpress-mobile/WordPressKit-iOS/pull/738 but in a neat and reviewable order:

- Move files into Sources/ and Tests/
- Introduce `Bundle` helper that differentiates between SPM and CocoaPods installations
- Isolate core API objects in dedicated package
- Then, new ground, define protocol abstraction to work around the Objective-C / Swift inter-op issue

See also, https://github.com/wordpress-mobile/WordPressKit-iOS/pull/756 which is related to the SPM work but independent from this one.

---

- [ ] Please check here if your pull request includes additional test coverage. — N.A.
- [x] I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary. — N.A.

mokagio committed:
Add tests for the `NSDate` extension implementing RFC3339

mokagio committed:
Convert `-(NSString *)WordPressComJSONString` method to Swift

mokagio committed:
Convert `rfc3339DateFormatter` method to Swift

mokagio committed:
Convert `dateWithWordPressComJSONString:` to Swift

This resulted in a change in the Swift API. I think this is acceptable
because the method was not used outside WordPressKit.

See:

- WordPressAuthenticator https://github.com/search?q=repo%3Awordpress-mobile%2FWordPressAuthenticator-iOS%20dateWithWordPressComJSONString&type=code
- WordPress-iOS https://github.com/search?q=repo%3Awordpress-mobile%2FWordPress-iOS+dateWithWordPressComJSONString&type=code

mokagio committed:
Delete unused `NSDate+WordPressJSON` Objective-C file

mokagio committed:
Make new Swift `rfc3339DateFormatter` a `static let` instead of `func`

It didn't work in Objective-C, but in Swift it's possible to do it and
it makes more sense to have it implemented as `static let` so it's
initialized only once.

mokagio commented on WordPressKit/NSDate+RFC3339.swift:
```diff
@@ -0,0 +1,38 @@
+    }
+}
+
+extension DateFormatter {
+
+    static let rfc3339Formatter: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "yyyy'-'MM'-'dd'T'HH':'mm':'ssZ"
+        formatter.timeZone = NSTimeZone(forSecondsFromGMT: 0) as TimeZone
+        formatter.locale = NSLocale(localeIdentifier: "en_US_POSIX") as Locale
+        return formatter
+    }()
+}
```
It seemed neater to define this pre-configured `DateFormatter` in a `DateFormatter` extension, rather than `NSDate`.

mokagio commented on WordPressKit/PostServiceRemoteREST+Extended.swift:
```diff
@@ -40,7 +40,7 @@ private func decodePost(from object: AnyObject) async throws -> RemotePost {
 private func makeParameters<T: Encodable>(from value: T) throws -> [String: AnyObject] {
     let encoder = JSONEncoder()
-    encoder.dateEncodingStrategy = .formatted(NSDate.rfc3339DateFormatter())
+    encoder.dateEncodingStrategy = .formatted(NSDate.rfc3339DateFormatter)
```
I forgot about the API breakage concerns when making this a `static let`. 🤦‍♂️ 

Turns out there is one use of this outside the library, [in the WordPress iOS tests](https://github.com/search?q=repo%3Awordpress-mobile%2FWordPress-iOS+rfc3339DateFormatter&type=code). 

I'll update the code reintroducing the `static func` but marking it as deprecated. This way, we can remove it at the next "serious" breaking change occasion.

mokagio committed:
Restore `rfc3339DateFormatter()` method for backwards compatibility

crazytonyli commented:
> Let me know if you think I'm being slack, or if it sounds like a reasonable compromise to move ahead without distractions.

My preference would be religiously following semantic versioning. There is no really difference in publishing 15.0 vs 14.2. Plus, when releasing a major version, you get to further clean up the codebase, i.e. moving `extension NSDate` to `extension Date`, removing `NSDate.rfcXxxDateFormatter` and using the one in `DateFormatter` instead, etc.

crazytonyli commented on WordPressKit/NSDate+RFC3339.swift:
```diff
@@ -0,0 +1,43 @@
+
+extension DateFormatter {
+
+    static let rfc3339Formatter: DateFormatter = {
```
If you are going to make breaking changes, I guess we have an option of getting rid of the `NSDate` extension above and renaming this `rfc3339Formatter` to be something like `wordPressComDateFormatter`.

crazytonyli commented on WordPressKitTests/NSDate+RFC3339Tests.swift:
```diff
@@ -0,0 +1,43 @@
+            NSDate(timeIntervalSince1970: 1_679_238_000).wordPressComJSONString(),
+            // Apparently, NSDateFormatter doesn't offer a way to specify Z vs +0000.
+            // This might go all the way back to the ISO 8601 and RFC 3339 specs overlap.
+            "2023-03-19T15:00:00+0000"
```
Have you tried to put a single quote `'` around the `Z` in the format: `"yyyy'-'MM'-'dd'T'HH':'mm':'ss'Z'"`?

crazytonyli replied:
BTW, this is one of the reasons I think we should avoid using standard names in function names. It's pretty difficult to ensure conformance.

crazytonyli replied:
If we are being really strict here, this test case should fail, because the output is not in [ RFC3339 time format](https://datatracker.ietf.org/doc/html/rfc3339#section-5.6).

mokagio replied:
Thanks @crazytonyli !

I didn't noticed that 🤦‍♂️ 

I tried locally and it does the job at returning `2023-03-19T15:00:00Z` as I originally expected. 

But ☝️ other tests that deal with API and dates fail. I think maybe we should leave it like this for the moment, and update / investigate later on. After all, the format the code uses right now is the "original" one.

What do you think?

mokagio commented:
> My preference would be religiously following semantic versioning

Good point. It also makes the decision making simpler "is this a breaking change (regardless of whether the API is `public` but with no consumers)" "yes?" "new major version."

Boom 🚀

mokagio committed:
Bump version to 15.0.0-beta.1 because of breaking change in `NSDate` API

See 71fa95137a6687c6f28ea18e78322de3622ed997

mokagio committed:
Breaking change – Remove deprecated `rfc3339DateFormatter()`

mokagio committed:
Extract `DateFormatter` RFC-3339 to file and rename to `wordPressCom`

mokagio committed:
Breaking change – Move `NSDate` WordPress.com implementation to `Date`

The breaking change is actually in having removed
`NSDate.rfc3339Formatter`, which is no longer necessary.

mokagio commented:
@crazytonyli thanks again for the suggestions. I implemented them in https://github.com/wordpress-mobile/WordPressKit-iOS/pull/759

mokagio committed:
Breaking change – Modernize `NSDate` and `Date` WP.com extensions (#759)

mokagio committed:
Add `CHANGELOG` entry for reworked WP.com `NSDate` conversion

crazytonyli commented on WordPressKit.podspec:
```diff
@@ -2,7 +2,7 @@
 Pod::Spec.new do |s|
   s.name          = 'WordPressKit'
-  s.version       = '14.1.0'
+  s.version       = '15.0.0-beta.1'
```
Usually we make version change in a dedicated PR along side with moving changelog content. Do you intend to make a release once this PR is merged?

mokagio replied:
Oh, right. I just wanted to put a line in the sand for the breaking change.
